compiler/aarch64/codegen/ControlFlowEvaluator.cpp

```diff
@@ -514,8 +515,70 @@ OMR::ARM64::TreeEvaluator::lookupEvaluator(TR::Node *node, TR::CodeGenerator *cg
 TR::Register *
 OMR::ARM64::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   // Only temporary (#3858 implements this)
-   cg->comp()->failCompilation<TR::AssertionFailure>("tableEvaluator");
+   int32_t numBranchTableEntries = node->getNumChildren() - 2;
+   TR::Node *defaultChild = node->getSecondChild();
+   TR::Register *selectorReg = cg->evaluate(node->getFirstChild());
+   TR::Register *tmpRegister = NULL;
+   TR::RegisterDependencyConditions *conditions;
+   int32_t i;
+
+   if (5 <= numBranchTableEntries)
+      {
+      conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(2, 2, cg->trMemory());
+      tmpRegister = cg->allocateRegister();
+      TR::addDependency(conditions, tmpRegister, TR::RealRegister::NoReg, TR_GPR, cg);
+      }
+   else
+      {
+      conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(1, 1, cg->trMemory());
+      }
+
+   TR::addDependency(conditions, selectorReg, TR::RealRegister::NoReg, TR_GPR, cg);
+
+   if (0 < defaultChild->getNumChildren())
+      {
+      cg->evaluate(defaultChild->getFirstChild());
+      conditions = conditions->clone(cg, generateRegisterDependencyConditions(cg, defaultChild->getFirstChild(), 0));
+      }
+
+   if (5 > numBranchTableEntries)
+      {
+      for (i = 0; i < numBranchTableEntries; i++)
+         {
+         generateCompareImmInstruction(cg, node, selectorReg, i);
+         generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, node->getChild(2+i)->getBranchDestination()->getNode()->getLabel(), TR::CC_EQ);
+         }
+
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, defaultChild->getBranchDestination()->getNode()->getLabel(), conditions);
+      }
+   else
+      {
+      if (!constantIsUnsignedImm12(numBranchTableEntries))
```

Comment:
I think this `if` condition is being overly defensive.  You already know from the `if` guarding entry to this code that the `numBranchTableEntries` is <= 5.  I'd say just do the `generateCompareImmInstruction`.

Reply:
`numBranchTableEntries` is >=5 here.  So it is possible that exceeds the range of unsigned imm12.

Reply:
Ugh.  I read the condition above incorrectly.